### PR TITLE
Add groundwork for tool management

### LINKED ISSUE #44: Management of locally installed dotnet tools
I definitely agree about installing tools being the path for getting this executor up and running. 

I see something like this:
1. Init schematic is updated to create the tool-manifest, with some messaging about "Tool Manifest created for managing dotnet tools"
2. Tools are installed if needed at the time a schematic that would use them is used the first time.
    * In practice, this means that we can have this format tool, or the swagger cli tool later, or other tools be known by the package, but only installed if users care about them.
    * A process flow with this would look something like: User runs `nx format my-api` -> Executor checks for tool in manifest, installs if missing -> Executor checks that tool has been restored, reports error message otherwise -> Executor runs tool
3. Add a restore schematic (simple schematic, runs `dotnet restore` and `dotnet tool restore`)
4. Update init schematic to add `nx g @nx-dotnet/core:restore` to npm prepare scripts. Something like:
    ```typescript
        const package = readJson(host, 'package.json')
        const prepareSteps = package.scripts.prepare?.split?.('&&')?.map?.(x => x.trim()) ?? [];
        if (!prepareSteps.includes('nx g @nx-dotnet/core:restore')) {
            prepareSteps.push('nx g @nx-dotnet/core:restore')
        }
        package.scripts.prepare = prepareSteps.join(' && ')
        writeJson(host, 'package.json', package)
    ```

I'll create new issues for the tool management when I am home this evening.

After we get tool-mgmt in, this executor should be pretty quick to test.

_Originally posted by @AgentEnder in https://github.com/nx-dotnet/nx-dotnet/issues/13#issuecomment-842633817_

### PR DESCRIPTION
**Pull request by bcallaghan-et**
Add the features and utilities for upcoming generators and executors that require .NET tools.

Fixes #44

### CODE REVIEW

File: packages/core/src/generators/init/generator.spec.ts [hunk 1]
```diff
@@ -1,33 +1,68 @@
...
+  it('should not add restore if it already exists', async () => {
+    const packageJson = {
+      scripts: { prepare: 'nx g @nx-dotnet/core:restore' },
+    };
+    writeJson(appTree, 'package.json', packageJson);
+    await generator(appTree, dotnetClient);
+    const updated = readJson(appTree, 'package.json');
+    expect(updated.scripts.prepare).toBe('nx g @nx-dotnet/core:restore');
+  });
 });
```
Comment (AgentEnder) [inline packages/core/src/generators/init/generator.spec.ts:68]: I'd recommend adding another test to ensure that everything works and the user's existing prepare scripts are not removed.

**Review by AgentEnder — COMMENTED** [review]
Looks good, add the 1 test and then I will test this out when I am home and we can get it merged in.

File: packages/core/src/generators/init/generator.spec.ts (hunk 1 above)
Comment (bcallaghan-et) [reply packages/core/src/generators/init/generator.spec.ts:80]: I've added the test you requested.

**Comment by AgentEnder** [comment]
@bcallaghan-et In testing this at home I ran into the init schematic failing when calling it via `nx g @nx-dotnet/core:init`, unless it had already been called by a different schematic (i.e. the app generator). This is because nx generators are **always** called with two parameters by the CLI. Host, and options.

Our init schematic currently does not take any options, but Nx still passes it.  Since we need to add a default parameter, we also have to add an unused second parameter (I recommend just using an `_` for it for now, we can type it and change its name to options later if we need them.

**Comment by AgentEnder** [comment]
The unit tests do not catch this since we are importing the generator and calling it directly. E2E tests should catch this though. I am working on getting the e2e in better shape, I do not like how it works currently.

File: packages/core/src/generators/init/generator.ts [hunk 2]
```diff
@@ -3,13 +3,18 @@ import {
...
   writeJson,
 } from '@nrwl/devkit';
+import { DotNetClient, dotnetFactory } from '@nx-dotnet/dotnet';
 
 import { CONFIG_FILE_PATH, NxDotnetConfig } from '@nx-dotnet/utils';
 
-export default async function (host: Tree) {
+export default async function (
+  host: Tree,
+  dotnetClient = new DotNetClient(dotnetFactory()),
```
Comment (AgentEnder) [inline packages/core/src/generators/init/generator.ts:16]: Add a new parameter here between host + dotnetClient. I recommend 
```
export default async function (
  host: Tree,
  _: null,
  dotnetClient = new DotNetClient(dotnetFactory()),
) {
```

Nx CLI will call the function with 2 parameters (and the default dotnetClient), so we have to allocate the first two parameters to Nx.

**Pull request merged**